### PR TITLE
Dettach node device from host without device driver

### PR DESCRIPTION
**Pull request by sannyshao**
Check the driver of device and remove the driver.
Then detach the node device and reattach it.

Signed-off-by: Jingjing Shao <jishao@redhat.com>

### CODE REVIEW

**Comment by sannyshao** [comment]
Check the support for python:

2to3-3 virsh_nodedev_detach_reattach.py
RefactoringTool: Skipping optional fixer: buffer
RefactoringTool: Skipping optional fixer: idioms
RefactoringTool: Skipping optional fixer: set_literal
RefactoringTool: Skipping optional fixer: ws_comma
RefactoringTool: No changes to virsh_nodedev_detach_reattach.py
RefactoringTool: Files that need to be modified:
RefactoringTool: virsh_nodedev_detach_reattach.py

**Comment by sannyshao** [comment]
avocado run  --vt-type libvirt  type_specific.io-github-autotest-libvirt.virsh.nodedev_detach_reattach.positive.normal_test
JOB ID     : 8e584fce7fce52f6ceb2200685fe97be1437290c
JOB LOG    : /root/avocado/job-results/job-2018-03-08T16.34-8e584fc/job.log
 (1/1) type_specific.io-github-autotest-libvirt.virsh.nodedev_detach_reattach.positive.normal_test: PASS (3.51 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0

avocado run  --vt-type libvirt  type_specific.io-github-autotest-libvirt.virsh.nodedev_detach_reattach.positive.acl_test
JOB ID     : 8bc885baf11aa8a73baf05cc2a52509378906ecb
JOB LOG    : /root/avocado/job-results/job-2018-03-08T16.36-8bc885b/job.log
 (1/1) type_specific.io-github-autotest-libvirt.virsh.nodedev_detach_reattach.positive.acl_test: PASS (20.38 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0

avocado run  --vt-type libvirt  type_specific.io-github-autotest-libvirt.virsh.nodedev_detach_reattach.positive.non_driver_test
JOB ID     : 0736684dbbd9cc3c95bb939bf2c1de41ecfcd2ec
JOB LOG    : /root/avocado/job-results/job-2018-03-08T16.36-0736684/job.log
 (1/1) type_specific.io-github-autotest-libvirt.virsh.nodedev_detach_reattach.positive.non_driver_test: PASS (4.72 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0

avocado run  --vt-type libvirt  type_specific.io-github-autotest-libvirt.virsh.nodedev_detach_reattach.negative.no_arg
JOB ID     : 5b31b943fc3ad7eaea1872a8bae347cacadda53b
JOB LOG    : /root/avocado/job-results/job-2018-03-08T16.39-5b31b94/job.log
 (1/1) type_specific.io-github-autotest-libvirt.virsh.nodedev_detach_reattach.negative.no_arg: PASS (1.98 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 5.56 s

avocado run  --vt-type libvirt  type_specific.io-github-autotest-libvirt.virsh.nodedev_detach_reattach.negative.name_unrecognized
JOB ID     : 791bdcb9896fc0a8fe32e7836b3df36e6120be3a
JOB LOG    : /root/avocado/job-results/job-2018-03-08T16.40-791bdcb/job.log
 (1/1) type_specific.io-github-autotest-libvirt.virsh.nodedev_detach_reattach.negative.name_unrecognized: PASS (2.02 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0

avocado run  --vt-type libvirt  type_specific.io-github-autotest-libvirt.virsh.nodedev_detach_reattach.negative.multi_detach
JOB ID     : 08eed14a8f5285c5b92fc1d24336df17cd8f11e1
JOB LOG    : /root/avocado/job-results/job-2018-03-08T16.41-08eed14/job.log
 (1/1) type_specific.io-github-autotest-libvirt.virsh.nodedev_detach_reattach.negative.multi_detach: PASS (2.02 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0

avocado run  --vt-type libvirt  type_specific.io-github-autotest-libvirt.virsh.nodedev_detach_reattach.negative.not_pci_device
JOB ID     : 7d3b60618da2afbcaf6fb5c0ed4fb8f80c67e183
JOB LOG    : /root/avocado/job-results/job-2018-03-08T16.41-7d3b606/job.log
 (1/1) type_specific.io-github-autotest-libvirt.virsh.nodedev_detach_reattach.negative.not_pci_device: PASS (1.96 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0

avocado run  --vt-type libvirt   type_specific.io-github-autotest-libvirt.virsh.nodedev_detach_reattach.negative.nodedev_unknowopt
JOB ID     : afd081d77d998d934dca4a07f19097f15e8dc7f4
JOB LOG    : /root/avocado/job-results/job-2018-03-08T16.42-afd081d/job.log
 (1/1) type_specific.io-github-autotest-libvirt.virsh.nodedev_detach_reattach.negative.nodedev_unknowopt: PASS (2.82 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0

avocado run  --vt-type libvirt  type_specific.io-github-autotest-libvirt.virsh.nodedev_detach_reattach.negative.acl_test
JOB ID     : e4c429cf5c2e26be67c8c9676aa5a9bd014ba3ce
JOB LOG    : /root/avocado/job-results/job-2018-03-08T16.42-e4c429c/job.log
 (1/1) type_specific.io-github-autotest-libvirt.virsh.nodedev_detach_reattach.negative.acl_test: PASS (50.00 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py [hunk 1]
```diff
@@ -116,16 +133,23 @@ def check_kernel_option():
...
+    check_driver = params.get('check_driver', '')
     # check variables.
     if not libvirt_version.version_compare(1, 1, 1):
         if params.get('setup_libvirt_polkit') == 'yes':
             test.cancel("API acl test not supported in current"
                         " libvirt version.")
 
+    # check the device driver and delete the driver
+    if not check_driver:
+        print "001"
```
Comment (balamuruhans) [inline libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: please remove this line

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py [hunk 2]
```diff
@@ -116,16 +133,23 @@ def check_kernel_option():
...
+        device_driver = get_device_driver(device_address)
+        remove_cmd = "modprobe -r %s" % device_driver
+        remove_opt = process.run(remove_cmd, shell=True)
```
Comment (balamuruhans) [inline libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: checking the status of modprobe command and raise would be a better way. please use process.system()

File: libvirt/tests/cfg/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.cfg [hunk 3]
```diff
@@ -6,6 +6,7 @@
     nodedev_device = 'ENTER.YOUR.PCI.DEVICE.TO.DETACH'
     #nodedev_device_opt: options for nodedev cmd.
     nodedev_device_opt = ""
+    check_driver = "CHECK_DEVICE_DRIVER"
```
Comment (balamuruhans) [inline libvirt/tests/cfg/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.cfg]: check_driver = "yes"

File: libvirt/tests/cfg/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.cfg [hunk 4]
```diff
@@ -17,6 +18,9 @@
                     action_lookup = "connect_driver:QEMU"
                     unprivileged_user = "EXAMPLE"
                     virsh_uri = "qemu:///system"
+                - non_driver_test:
+                    check_driver = ''
```
Comment (balamuruhans) [inline libvirt/tests/cfg/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.cfg]: check_driver = "no"

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py [hunk 5]
```diff
@@ -116,16 +133,23 @@ def check_kernel_option():
             test.cancel('Param device_address is not configured.')
     device_opt = params.get('nodedev_device_opt', '')
     status_error = ('yes' == params.get('status_error', 'no'))
+    check_driver = params.get('check_driver', '')
```
Comment (balamuruhans) [inline libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: `check_driver = params.get('check_driver', 'no') == 'yes'`

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py [hunk 6]
```diff
@@ -136,3 +160,8 @@ def check_kernel_option():
     # Do nodedev detach and reattach success.
     if status_error:
         test.fail('Test successed in negative case.')
+
+    # reload the device driver
+    if not check_driver:
+        reload_cmd = "modprobe %s" % device_driver
+        reload_opt = process.run(reload_cmd, shell=True)
```
Comment (balamuruhans) [inline libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: same here

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py [hunk 7]
```diff
@@ -82,13 +91,21 @@ def pci_device_address():
         net_list = virsh.nodedev_list(tree='', cap='net')
         net_lists = net_list.stdout.strip().splitlines()
         route_cmd = " route | grep default"
-        route_default = process.run(route_cmd, shell=True).stdout_text.strip().split(' ')
+        route_default = process.run(route_cmd, shell=True).stdout.strip().split(' ')
```
Comment (balamuruhans) [inline libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: process.system_output() would be more appropriate, it takes stdout and strips()

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py [hunk 8]
```diff
@@ -116,16 +133,22 @@ def check_kernel_option():
...
     if not libvirt_version.version_compare(1, 1, 1):
         if params.get('setup_libvirt_polkit') == 'yes':
             test.cancel("API acl test not supported in current"
                         " libvirt version.")
 
+    # check the device driver and delete the driver
+    if not with_driver:
+        device_driver = get_device_driver(device_address)
+        remove_cmd = "modprobe -r %s" % device_driver
+        remove_opt = process.system(remove_cmd, shell=True)
```
Comment (balamuruhans) [inline libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: so to confirm whether driver is removed, we need to check the status here,
```python
if process.system(remove_cmd, shell=True) !=0:
    test.error()
```

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py [hunk 9]
```diff
@@ -136,3 +159,8 @@ def check_kernel_option():
     # Do nodedev detach and reattach success.
     if status_error:
         test.fail('Test successed in negative case.')
+
+    # reload the device driver
+    if not with_driver:
+        reload_cmd = "modprobe %s" % device_driver
+        reload_opt = process.system(reload_cmd, shell=True)
```
Comment (balamuruhans) [inline libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: same check to be considered here.

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py [hunk 10]
```diff
@@ -2,6 +2,7 @@
 import logging
 from virttest import virsh
 from virttest.libvirt_xml import nodedev_xml
+
```
Comment (dzhengfy) [inline libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: Remove this blank line, please.

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py [hunk 11]
```diff
@@ -28,6 +29,14 @@ def get_driver_readlink(device_address):
             return None
         return driver
 
+    def get_device_driver(device_address):
+        """
+        get the driver of device.
```
Comment (dzhengfy) [inline libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: Capital the first letter, please.

Comment (dzhengfy) [reply libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: Please add docstring for parameter, like 
:param device_address: xxxxx

:return: xxx

File: libvirt/tests/cfg/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.cfg [hunk 12]
```diff
@@ -17,6 +18,9 @@
                     action_lookup = "connect_driver:QEMU"
                     unprivileged_user = "EXAMPLE"
                     virsh_uri = "qemu:///system"
+                - non_driver_test:
+                    with_driver = 'no' 
+
```
[the remark this answers is not in the document]
Comment (sannyshao) [reply libvirt/tests/cfg/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.cfg]: Delete it， done

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py (hunk 10 above)
Comment (sannyshao) [reply libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: done

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py (hunk 9 above)
Comment (sannyshao) [reply libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: done

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py (hunk 8 above)
Comment (sannyshao) [reply libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: done

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py (hunk 11 above)
Comment (sannyshao) [reply libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: done

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (sannyshao) [reply libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: done

**Pull request merged**